Reviews table

chloealee opened the pull request:
#### What's this PR do?
- When a photo post is reviewed, a review is added to the reviews table. 
- Updates `Reviews` table to include `postable_id` and `postable_type`.
- Updates relations between reviews and posts.
- Updates documentation (now this endpoint requires a `reviewer` and `event_type`).

#### How should this be reviewed?
- Post a review to the `/reviews` endpoint. 
  - Make sure an event with the  correct information is created.
  - Make sure a review with the correct information is created.

#### Relevant tickets
Fixes #117 
Fixes #113 

#### Checklist
- [x] Documentation added for new features/changed endpoints.
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

chloealee commented on app/Repositories/PhotoRepository.php:
```diff
@@ -154,12 +155,44 @@ public function reviews($data)
+                    $event = Event::create([
+                        'signup_id' => $post->signup_id,
+                        // Do we want the user's northstar id or the admin's?
+                        'northstar_id' => $post->northstar_id,
```
@sbsmith86 @katiecrane @DFurnes do we want the user's northstar id or the admin's? Right now, this is the user's but I'm thinking maybe it makes more sense to have the admin's?

chloealee commented on app/Repositories/ReportbackRepository.php:
```diff
@@ -179,11 +181,48 @@ public function updateReportbackItems($data)
+                    $event = Event::create([
+                        'signup_id' => 0,
+                        // Do we want the user's northstar id or the admin's?
+                        'northstar_id' => $rb->northstar_id,
```
@sbsmith86 @katiecrane @DFurnes same with this! Should we change it to the admin's northstar id?

DFurnes commented on app/Repositories/ReportbackRepository.php:
```diff
@@ -179,11 +181,47 @@ public function updateReportbackItems($data)
         foreach ($data as $reportbackItem) {
```
Do we want to allow mass-reviewing reportbacks? It doesn't feel super RESTful (which isn't the end of the world), and it also seems like we might be working ourselves into the same corner we were in in Phoenix (where multiple reviewers couldn't safely be working on the same campaign without conflicts).

sbsmith86 replied:
@DFurnes in the new interface, you would just review one at a time, but in Phoenix you review a bunch and then submit. This is one of the reasons why I brought up what I did in Slack. I would like to not have to support the Phoenix interface at all in Rogue.

ngjo replied:
yes to what dave said! in rogue we'd want multiple reviewers to be able to safely work on the same campaign w/o conflicts

sbsmith86 replied:
@chloealee yea i think we shouldn't allow this.

sbsmith86 commented on app/Models/ReportbackItem.php:
```diff
@@ -28,4 +28,12 @@ public function reactions()
+    /**
+     * Each reportback item has one review.
+     */
+    public function review()
```
@DFurnes yea they are. ReportbackItem was from the old DB structure and `Post` is the new. At one point we had reviews coming in from both so we had to continue to support both structures which is one of the reasons we turned off sending to rogue so that we could just get all of the reportbacks into rogue as a `Post`. So I don't think this PR needs to support `ReportbackItem`s anymore @chloealee

chloealee replied:
@sbsmith86 ah totally right I wasn't thinking - will update now!

chloealee commented:
@DFurnes @sbsmith86 @katiecrane finally finished with the unit tests. This is ready for another 👀 !

chloealee commented:
Right now, Phoenix no longer sends over the `reviewer` to this endpoint but this PR now requires it. We don't have to update Phoenix since it'll never use this endpoint anymore, right? cc @DFurnes @sbsmith86 @katiecrane

katiecrane commented on app/Http/Controllers/Api/ReviewsController.php:
```diff
@@ -76,13 +77,12 @@ public function reviews(Request $request)
             } else {
                 $review['rogue_reportback_item_id'] = $review['rogue_event_id'];
-                array_push($reportbackItems, array_only($review, ['rogue_reportback_item_id', 'status']));
+                array_push($reportbackItems, array_except($review, ['rogue_event_id']));
```
@chloealee is this still used ever?

chloealee replied:
@katiecrane after our conversation, I guess it won't be since we won't ever be reviewing in Phoenix and sending to Rogue! Should I delete all the reportback item logic?

katiecrane replied:
@chloealee my understanding is that we can get rid of all of that

sbsmith86 commented on app/Repositories/PhotoRepository.php:
```diff
@@ -154,17 +155,48 @@ public function reviews($data)
-
-                    array_push($reviewedPhotos, $post->content);
+                    // Create the Event.
+                    $event = Event::create([
```
should we hold off on this until we do the work of moving the even creation logic into a model event? Or we can move forward and fix later.

chloealee replied:
@sbsmith86 I don't feel strongly either way! Maybe keep for now since there is a test to make sure an event is created and we'll want to keep this test even when we switch to creation logic into a model event?

sbsmith86 commented on app/Repositories/ReportbackRepository.php:
```diff
@@ -179,11 +179,14 @@ public function updateReportbackItems($data)
         foreach ($data as $reportbackItem) {
             if (isset($reportbackItem['rogue_reportback_item_id']) && ! empty($reportbackItem['rogue_reportback_item_id'])) {
```
And we don't need to support reportback item ids anymore.

chloealee replied:
@sbsmith86 @katiecrane removed all the added reportback item logic!

katiecrane commented on app/Http/Controllers/Api/ReviewsController.php:
```diff
@@ -67,35 +67,21 @@ public function reviews(Request $request)
-            $reviewedReportbackItemsCode = $this->code($reviewedReportbackItems);
-        }
-
         if ($photos) {
```
@chloealee is this always true?

sbsmith86 commented on tests/ReviewsApiTest.php:
```diff
@@ -0,0 +1,142 @@
+     * PUT /reviews
+     * @return void
+     */
+    public function testUpdatingMultipleReviews()
```
So I'm not sure this is functionality we really want. In the new interface, an admin would hit approve/reject one at a time instead of how it works now where you change a bunch of statuses and then submit them. We storta talked about it in a comment chain before. I don't know if it is definitely need to merge this PR but we should only expect to be reviewing posts one at a time.

chloealee replied:
ah that's right - sorry! Should I go ahead and update that here @sbsmith86 ?

chloealee replied:
I may as well - I'll do that now!

katiecrane commented on app/Repositories/ReportbackRepository.php:
```diff
@@ -179,14 +179,9 @@ public function updateReportbackItems($data)
         foreach ($data as $reportbackItem) {
             if (isset($reportbackItem['rogue_reportback_item_id']) && ! empty($reportbackItem['rogue_reportback_item_id'])) {
                 $rbItem = ReportbackItem::where(['id' => $reportbackItem['rogue_reportback_item_id']])->first();
```
@chloealee does this part get used?

chloealee replied:
@katiecrane this is old code!

katiecrane replied:
@chloealee I don't know how I did this so sorry

chloealee commented on app/Http/Controllers/Api/ReviewsController.php:
```diff
@@ -55,48 +55,19 @@ public function __construct(PostService $posts, ReportbackService $reportbacks)
-        $this->validate($request, [
-            '*.rogue_event_id' => 'required',
-            '*.status' => 'required',
-        ]);
```
@sbsmith86 @katiecrane moved this out to it's own `ReviewsRequest.php` since we aren't dealing with multiple reportback items anymore. 

@sbsmith86 I got it to work (you were right, my headers were wrong in Postman!) to return a `500` response code. However, the message is not filled in automatically like it is when I don't follow validation rules in PostRequest. I also didn't follow validation rules in the ReactionRequest and no message is returned there either, just a `500` code. Is this ok to just return a `500` or should it be also returning the message?

chloealee commented on app/Http/Requests/ReviewsRequest.php:
```diff
@@ -0,0 +1,31 @@
+<?php
```
@sbsmith86 @katiecrane here is the new validation (instead of in the Reivews Controller!)

sbsmith86 commented on app/Http/Controllers/Api/PostsController.php:
```diff
@@ -121,6 +116,6 @@ public function reviews(Request $request)
 
         // @TODO: we'll need to change the transformer here depending on what type of post.
-        return $this->collection($reviewedPosts, $code, $meta, $this->photoTransformer);
+        return $this->item($reviewedPost, $code, $meta, $this->photoTransformer);
```
We have the `PostTransformer` that should be used across all post types.

chloealee replied:
@sbsmith86 I actually don't think we even need this here? I don't think it's being used!

chloealee replied:
wait sorry - I am totally confusing myself. It should go through to the Post Controller but it looks like it skips and goes right to the Post Service. I will fix this!

chloealee replied:
@sbsmith86 or is it ok for it to go straight to the Post Service?

sbsmith86 commented on tests/ReviewsApiTest.php:
```diff
@@ -0,0 +1,62 @@
+        // Create a photo.
+        $photo = factory(Photo::class)->create();
+
+        // Create a post and associate the photo with it.
```
I wonder if we should move this work into a model event also. So when a type of thing is created we also automatically create the post for it. Not needed for merge, just a thought.

sbsmith86 commented:
Looks good! I just had one comment about using the `PostTransformer`instead of the `PhotoTransformer`

chloealee commented:
Awesome, thanks @sbsmith86 ! Just took out the photo transformer and replaced with the post transformer (thanks for that!) and updated documentation.

sbsmith86 commented on app/Http/Controllers/Api/ReviewsController.php:
```diff
@@ -3,9 +3,9 @@
 use Rogue\Services\PostService;
 use Rogue\Services\ReportbackService;
+use Rogue\Http\Requests\ReviewsRequest;
 use Rogue\Http\Transformers\PhotoTransformer;
```
Sorrrrry 😬  I think you can use the `PostTransformer` here to?

chloealee commented:
thanks for being so patient @sbsmith86 ! Just updated so it returns a post and uses the post transformer and also updated the test and documentation to reflect this!
